sdk/core/azure-core-amqp/src/test/java/com/azure/core/amqp/implementation/MessageFluxIsolatedTest.java

```diff
@@ -0,0 +1,909 @@
+        when(firstReceiver.receive()).thenReturn(firstReceiverFacade.getMessages());
+        when(firstReceiver.closeAsync()).thenReturn(Mono.empty());
+
+        final ReactorReceiver secondReceiver = mock(ReactorReceiver.class);
+        final ReactorReceiverFacade secondReceiverFacade = new ReactorReceiverFacade(upstream, secondReceiver);
+        when(secondReceiver.getEndpointStates()).thenReturn(secondReceiverFacade.getEndpointStates());
+        when(secondReceiver.receive()).thenReturn(secondReceiverFacade.getMessages());
+        when(secondReceiver.closeAsync()).thenReturn(Mono.empty());
+
+        try (VirtualTimeStepVerifier verifier = new VirtualTimeStepVerifier()) {
+            verifier.create(() -> messageFlux)
+                .then(firstReceiverFacade.emit())
+                .then(firstReceiverFacade.errorEndpointStates(new AmqpException(false, "non-retriable", null)))
+                .thenAwait(UPSTREAM_DELAY_BEFORE_NEXT)
+                .then(() -> firstReceiverFacade.assertNoPendingSubscriptionsToMessages())
+                .then(secondReceiverFacade.emit())
+                .then(() -> upstream.complete())
+                .verifyError();
+        }
+
+        Assertions.assertTrue(firstReceiverFacade.wasSubscribedToMessages());
+        Assertions.assertFalse(secondReceiverFacade.wasSubscribedToMessages());
+        // Expecting closeAsync invocation from two call sites -
+        // 1. after identified that receiver is terminated.
+        // 2. when operator terminates due to non-retriable error.
+        verify(firstReceiver, new AtLeast(2)).closeAsync();
+        verify(secondReceiver, never()).closeAsync();
+        upstream.assertCancelled();
+    }
+
+    @ParameterizedTest
+    @CsvSource({
+        "EmissionDriven,1",
+        "RequestDriven,0"
+    })
+    @Execution(ExecutionMode.SAME_THREAD)
+    public void shouldTerminateWhenRetriesOfReceiversErrorExhausts(CreditFlowMode creditFlowMode, int prefetch) {
+        final AmqpException error = new AmqpException(true, "retriable", null);
+        final TestPublisher<ReactorReceiver> upstream = TestPublisher.create();
+        final MessageFlux messageFlux = new MessageFlux(upstream.flux(), prefetch, creditFlowMode, retryPolicy);
+
+        final ReactorReceiver receiver = mock(ReactorReceiver.class);
+        final ReactorReceiverFacade receiverFacade = new ReactorReceiverFacade(upstream,
+            receiver,
+            TestPublisher.<Message>createCold());
+        when(receiver.getEndpointStates()).thenReturn(receiverFacade.getEndpointStates());
+        when(receiver.receive()).thenReturn(receiverFacade.getMessages());
+        when(receiver.closeAsync()).thenReturn(Mono.empty());
+
+        try (VirtualTimeStepVerifier verifier = new VirtualTimeStepVerifier()) {
+            verifier.create(() -> messageFlux)
+                .then(receiverFacade.emit())
```

Comment:
This might be easier to read if you wrap these in a single runnable? Or does it make a big difference to split them into multiple calls?  It looked overwhelming when I first saw it.
```java
.then(() -> {
   receiverFacade.emit();
   receiverFacade..errorEndpointStates(error);
   receiverFacade.completeMessages();
})
.thenAwait()
```

Reply:
Hmm… since all these methods in Façade type returns `Runnable`, putting these methods it in single `then` in some places means, we need to put `.run` on each of methods. Which seems to complicate the readability.